File: android/src/main/java/app/plink/android/protocol/BaselineJpegValidator.kt
package app.plink.android.protocol

/** Validates complete baseline scans without allocating or reconstructing pixels. */
internal class BaselineJpegValidator(private val bytes: ByteArray) {
    private data class Component(val horizontal: Int, val vertical: Int, val quantization: Int)
    private data class Table(val counts: List<Int>, val symbols: List<Int>)
    private var offset = 2
    private val components = mutableMapOf<Int, Component>()
    private val scanned = mutableSetOf<Int>()
    private val tables = mutableMapOf<Int, Table>()
    private val quantizationTables = mutableSetOf<Int>()
    private var restartInterval = 0
    private var bitCount = 0
    private var bitByte = 0

    fun validate(width: Int, height: Int) {
        require(width > 0 && height > 0 &&
            maxOf(width, height) <= ScreenPreviewPayloadPolicy.maxLongEdge &&
            minOf(width, height) <= ScreenPreviewPayloadPolicy.maxShortEdge &&
            width <= ScreenPreviewPayloadPolicy.maxPixels / height &&
            bytes.size in 10..ScreenPreviewPayloadPolicy.maxJpegBytes &&
            byte(0) == 0xff && byte(1) == 0xd8)
        while (offset < bytes.size) {
            val marker = marker()
            if (marker == 0xd9) {
                require(components.isNotEmpty() && scanned.size == components.size && offset == bytes.size)
                return
            }
            require(offset + 2 <= bytes.size)
            val length = word(offset)
            val start = offset + 2
            val end = offset + length
            require(length >= 2 && end <= bytes.size)
            offset = end
            when (marker) {
                0xc0 -> {
                    require(components.isEmpty() && length >= 8 && byte(start) == 8 &&
                        word(start + 1) == height && word(start + 3) == width)
                    val count = byte(start + 5)
                    require(count in 1..4 && length == 8 + 3 * count)
                    repeat(count) { i ->
                        val p = start + 6 + 3 * i
                        val id = byte(p)
                        val h = byte(p + 1) shr 4
                        val v = byte(p + 1) and 15
                        require(id !in components && h in 1..4 && v in 1..4 && byte(p + 2) <= 3)
                        components[id] = Component(h, v, byte(p + 2))
                    }
                }
                0xc4 -> readTables(start, end)
                0xdb -> readQuantization(start, end)
                0xdd -> { require(length == 4); restartInterval = word(start) }
                0xda -> readScan(start, end, width, height)
                0xe1 -> validateExif(start, end)
                in 0xe0..0xef, 0xfe -> Unit
                else -> error("Unsupported JPEG marker.")
            }
        }
        error("Incomplete JPEG.")
    }

    private fun byte(position: Int): Int = bytes[position].toInt() and 0xff
    private fun word(position: Int): Int = (byte(position) shl 8) or byte(position + 1)

    private fun marker(): Int {
        require(offset < bytes.size && byte(offset) == 0xff)
        while (offset < bytes.size && byte(offset) == 0xff) offset++
        require(offset < bytes.size && byte(offset) != 0)
        return byte(offset++)
    }

    private fun readTables(start: Int, end: Int) {
        var p = start
        require(p < end)
        while (p < end) {
            require(p + 17 <= end)
            val id = byte(p)
            require(id shr 4 <= 1 && id and 15 <= 3)
            val counts = (p + 1 until p + 17).map(::byte)
            val count = counts.sum()
            p += 17
            require(count in 1..256 && p + count <= end)
            var code = 0
            counts.forEachIndexed { index, n ->
                code += n
                // All-ones codes are forbidden; padding cannot supply missing blocks.
                require(code < (1 shl (index + 1)))
                code = code shl 1
            }
            val symbols = (p until p + count).map(::byte)
            for (symbol in symbols) {
                require(if (id shr 4 == 0) symbol <= 11
                    else symbol == 0 || symbol == 0xf0 || symbol and 15 in 1..10)
            }
            tables[id] = Table(counts, symbols)
            p += count
        }
    }

    private fun readQuantization(start: Int, end: Int) {
        var p = start
        require(p < end)
        while (p < end) {
            val id = byte(p++)
            // Baseline sequential JPEG uses 8-bit, nonzero quantization entries.
            require(id in 0..3 && p + 64 <= end)
            repeat(64) { require(byte(p++) != 0) }
            quantizationTables += id
        }
    }

    private fun readScan(start: Int, end: Int, width: Int, height: Int) {
        require(components.isNotEmpty() && start < end)
        val count = byte(start)
        require(count in 1..4 && end - start == 4 + 2 * count &&
            byte(end - 3) == 0 && byte(end - 2) == 63 && byte(end - 1) == 0)
        val maxH = components.values.maxOf { it.horizontal }
        val maxV = components.values.maxOf { it.vertical }
        val blocks = mutableListOf<Pair<Table, Table>>()
        val scanIds = mutableSetOf<Int>()
        var columns = (width + 8 * maxH - 1) / (8 * maxH)
        var rows = (height + 8 * maxV - 1) / (8 * maxV)
        repeat(count) { i ->
            val id = byte(start + 1 + 2 * i)
            val selector = byte(start + 2 + 2 * i)
            val component = requireNotNull(components[id])
            require(component.quantization in quantizationTables)
            require(id !in scanned && scanIds.add(id))
            val dc = requireNotNull(tables[selector shr 4])
            val ac = requireNotNull(tables[0x10 or (selector and 15)])
            if (count == 1) {
                columns = (width * component.horizontal + 8 * maxH - 1) / (8 * maxH)
                rows = (height * component.vertical + 8 * maxV - 1) / (8 * maxV)
            }
            repeat(if (count == 1) 1 else component.horizontal * component.vertical) {
                blocks += dc to ac
            }
        }
        require(blocks.size <= 10)
        var restart = 0
        repeat(columns * rows) { mcu ->
            if (restartInterval > 0 && mcu > 0 && mcu % restartInterval == 0) {
                finishBits()
                require(marker() == 0xd0 + restart)
                restart = (restart + 1) % 8
            }
            for ((dc, ac) in blocks) {
                skipBits(symbol(dc))
                var coefficient = 1
                while (coefficient < 64) {
                    val value = symbol(ac)
                    if (value == 0) break
                    if (value == 0xf0) {
                        coefficient += 16
                        require(coefficient <= 64)
                    } else {
                        coefficient += value shr 4
                        require(coefficient < 64)
                        skipBits(value and 15)
                        coefficient++
                    }
                }
            }
        }
        finishBits()
        scanned += scanIds
    }

    private fun symbol(table: Table): Int {
        var code = 0
        var firstCode = 0
        var firstSymbol = 0
        for (count in table.counts) {
            code = (code shl 1) or readBit()
            val index = code - firstCode
            if (index in 0 until count) return table.symbols[firstSymbol + index]
            firstSymbol += count
            firstCode = (firstCode + count) shl 1
        }
        error("Invalid JPEG entropy code.")
    }

    private fun skipBits(count: Int) { repeat(count) { readBit() } }

    private fun readBit(): Int {
        if (bitCount == 0) {
            require(offset < bytes.size)
            bitByte = byte(offset++)
            if (bitByte == 0xff) {
                require(offset < bytes.size && byte(offset) == 0)
                offset++
            }
            bitCount = 8
        }
        bitCount--
        return (bitByte shr bitCount) and 1
    }

    private fun finishBits() {
        val mask = (1 shl bitCount) - 1
        require(bitByte and mask == mask)
        bitCount = 0
    }

    /** Only primary-image orientation applies; never follow thumbnail/GPS offsets. */
    private fun validateExif(start: Int, end: Int) {
        val signature = byteArrayOf(0x45, 0x78, 0x69, 0x66, 0, 0)
        if (end - start < 6 || signature.indices.any { bytes[start + it] != signature[it] }) return
        val base = start + 6
        require(end - base >= 8)
        val little = byte(base) == 0x49 && byte(base + 1) == 0x49
        require(little || byte(base) == 0x4d && byte(base + 1) == 0x4d)
        fun number(p: Int, size: Int): Long {
            require(p >= base && p <= end - size)
            var value = 0L
            repeat(size) { i ->
                value = value or (byte(p + i).toLong() shl (8 * if (little) i else size - i - 1))
            }
            return value
        }
        require(number(base + 2, 2) == 42L)
        val relative = number(base + 4, 4)
        require(relative in 8L..(end - base - 2).toLong())
        val ifd = base + relative.toInt()
        val count = number(ifd, 2).toInt()
        require(count <= (end - ifd - 2) / 12)
        var sawOrientation = false
        repeat(count) { i ->
            val p = ifd + 2 + 12 * i
            if (number(p, 2) == 0x112L) {
                require(!sawOrientation && number(p + 2, 2) == 3L &&
                    number(p + 4, 4) == 1L && number(p + 8, 2) == 1L)
                sawOrientation = true
            }
        }
    }
}

File: macos/Sources/PlinkCore/BaselineJPEGValidator.swift
import Foundation

/// Checks entropy completeness without reconstructing pixels. ImageIO can conceal
/// a truncated scan even when it reports statusComplete and the file ends in EOI.
struct BaselineJPEGValidator {
    private struct Component {
        let horizontal: Int
        let vertical: Int
        let quantization: UInt8
    }

    private struct HuffmanTable {
        let counts: [Int]
        let symbols: [UInt8]
    }

    private let bytes: [UInt8]
    private var offset = 2
    private var components: [UInt8: Component] = [:]
    private var scanned: Set<UInt8> = []
    private var tables: [UInt8: HuffmanTable] = [:]
    private var quantizationTables: Set<UInt8> = []
    private var restartInterval = 0
    private var bitCount = 0
    private var bitByte = 0

    init(data: Data) {
        bytes = [UInt8](data)
    }

    mutating func validate(width: Int, height: Int) throws {
        guard width > 0, height > 0,
              max(width, height) <= ScreenPreviewProtocol.maxLongEdge,
              min(width, height) <= ScreenPreviewProtocol.maxShortEdge,
              width <= ScreenPreviewProtocol.maxPixels / height,
              (10...ScreenPreviewProtocol.maxJPEGBytes).contains(bytes.count),
              bytes[0] == 0xff, bytes[1] == 0xd8
        else { throw PayloadPolicyError.malformedFrame }

        while offset < bytes.count {
            let marker = try readMarker()
            if marker == 0xd9 {
                guard !components.isEmpty, scanned.count == components.count,
                      offset == bytes.count else { throw PayloadPolicyError.malformedFrame }
                return
            }
            guard offset + 2 <= bytes.count else { throw PayloadPolicyError.malformedFrame }
            let length = word(at: offset)
            let start = offset + 2
            let end = offset + length
            guard length >= 2, end <= bytes.count else { throw PayloadPolicyError.malformedFrame }
            offset = end

            switch marker {
            case 0xc0:
                guard components.isEmpty, length >= 8, bytes[start] == 8,
                      word(at: start + 1) == height, word(at: start + 3) == width
                else { throw PayloadPolicyError.malformedFrame }
                let count = Int(bytes[start + 5])
                guard count > 0, length == 8 + 3 * count else { throw PayloadPolicyError.malformedFrame }
                for i in 0..<count {
                    let p = start + 6 + 3 * i
                    let id = bytes[p]
                    let h = Int(bytes[p + 1] >> 4)
                    let v = Int(bytes[p + 1] & 15)
                    guard components[id] == nil, (1...4).contains(h), (1...4).contains(v),
                          bytes[p + 2] <= 3 else { throw PayloadPolicyError.malformedFrame }
                    components[id] = Component(horizontal: h, vertical: v, quantization: bytes[p + 2])
                }
            case 0xc4:
                try readTables(start: start, end: end)
            case 0xdb:
                try readQuantization(start: start, end: end)
            case 0xdd:
                guard length == 4 else { throw PayloadPolicyError.malformedFrame }
                restartInterval = word(at: start)
            case 0xda:
                try readScan(start: start, end: end, width: width, height: height)
            case 0xe0...0xef, 0xfe:
                // Metadata is interpreted by ImageIO.
                break
            default:
                // Includes progressive, arithmetic, lossless, and stray restart markers.
                throw PayloadPolicyError.malformedFrame
            }
        }
        throw PayloadPolicyError.malformedFrame
    }

    private func word(at position: Int) -> Int {
        Int(bytes[position]) << 8 | Int(bytes[position + 1])
    }

    private mutating func readMarker() throws -> UInt8 {
        guard offset < bytes.count, bytes[offset] == 0xff else { throw PayloadPolicyError.malformedFrame }
        while offset < bytes.count && bytes[offset] == 0xff { offset += 1 }
        guard offset < bytes.count, bytes[offset] != 0 else { throw PayloadPolicyError.malformedFrame }
        defer { offset += 1 }
        return bytes[offset]
    }

    private mutating func readTables(start: Int, end: Int) throws {
        var p = start
        guard p < end else { throw PayloadPolicyError.malformedFrame }
        while p < end {
            guard p + 17 <= end else { throw PayloadPolicyError.malformedFrame }
            let id = bytes[p]
            guard id >> 4 <= 1, id & 15 <= 3 else { throw PayloadPolicyError.malformedFrame }
            let counts = bytes[(p + 1)..<(p + 17)].map(Int.init)
            let count = counts.reduce(0, +)
            p += 17
            guard count > 0, count <= 256, p + count <= end else { throw PayloadPolicyError.malformedFrame }
            // Canonical JPEG codes must leave the all-ones code unused. This also
            // prevents end-of-scan padding from supplying a missing block.
            var code = 0
            for (index, n) in counts.enumerated() {
                code += n
                guard code < (1 << (index + 1)) else { throw PayloadPolicyError.malformedFrame }
                code <<= 1
            }
            let symbols = Array(bytes[p..<(p + count)])
            for symbol in symbols {
                if id >> 4 == 0 {
                    guard symbol <= 11 else { throw PayloadPolicyError.malformedFrame }
                } else {
                    guard symbol == 0 || symbol == 0xf0 || (1...10).contains(Int(symbol & 15))
                    else { throw PayloadPolicyError.malformedFrame }
                }
            }
            tables[id] = HuffmanTable(counts: counts, symbols: symbols)
            p += count
        }
    }

    private mutating func readQuantization(start: Int, end: Int) throws {
        var p = start
        guard p < end else { throw PayloadPolicyError.malformedFrame }
        while p < end {
            let id = bytes[p]
            p += 1
            guard id <= 3, p + 64 <= end else { throw PayloadPolicyError.malformedFrame }
            for _ in 0..<64 {
                guard bytes[p] != 0 else { throw PayloadPolicyError.malformedFrame }
                p += 1
            }
            quantizationTables.insert(id)
        }
    }

    private mutating func readScan(start: Int, end: Int, width: Int, height: Int) throws {
        guard !components.isEmpty, start < end else { throw PayloadPolicyError.malformedFrame }
        let count = Int(bytes[start])
        guard (1...4).contains(count), end - start == 4 + 2 * count,
              bytes[end - 3] == 0, bytes[end - 2] == 63, bytes[end - 1] == 0
        else { throw PayloadPolicyError.malformedFrame }
        let maxH = components.values.map(\.horizontal).max()!
        let maxV = components.values.map(\.vertical).max()!
        var blocks: [(dc: HuffmanTable, ac: HuffmanTable)] = []
        var scanIDs: Set<UInt8> = []
        var columns = (width + 8 * maxH - 1) / (8 * maxH)
        var rows = (height + 8 * maxV - 1) / (8 * maxV)
        for i in 0..<count {
            let id = bytes[start + 1 + 2 * i]
            let selector = bytes[start + 2 + 2 * i]
            guard let component = components[id], !scanned.contains(id), scanIDs.insert(id).inserted,
                  quantizationTables.contains(component.quantization),
                  let dc = tables[selector >> 4], let ac = tables[0x10 | (selector & 15)]
            else { throw PayloadPolicyError.malformedFrame }
            if count == 1 {
                // Noninterleaved scans have one block per MCU, with no dummy edge blocks.
                columns = (width * component.horizontal + 8 * maxH - 1) / (8 * maxH)
                rows = (height * component.vertical + 8 * maxV - 1) / (8 * maxV)
            }
            let repetitions = count == 1 ? 1 : component.horizontal * component.vertical
            for _ in 0..<repetitions { blocks.append((dc, ac)) }
        }
        guard blocks.count <= 10 else { throw PayloadPolicyError.malformedFrame }
        var restart = 0
        for mcu in 0..<(columns * rows) {
            if restartInterval > 0, mcu > 0, mcu % restartInterval == 0 {
                try finishBits()
                guard try readMarker() == UInt8(0xd0 + restart) else { throw PayloadPolicyError.malformedFrame }
                restart = (restart + 1) % 8
            }
            for block in blocks {
                let dcSize = try symbol(using: block.dc)
                try skipBits(dcSize)
                var coefficient = 1
                while coefficient < 64 {
                    let ac = try symbol(using: block.ac)
                    if ac == 0 { break }
                    if ac == 0xf0 {
                        coefficient += 16
                        guard coefficient <= 64 else { throw PayloadPolicyError.malformedFrame }
                    } else {
                        coefficient += ac >> 4
                        guard coefficient < 64 else { throw PayloadPolicyError.malformedFrame }
                        try skipBits(ac & 15)
                        coefficient += 1
                    }
                }
            }
        }
        try finishBits()
        scanned.formUnion(scanIDs)
        // The outer parser must find a marker immediately after the final block.
        // It rejects extra entropy bytes, premature EOI, and trailing data.
    }

    private mutating func symbol(using table: HuffmanTable) throws -> Int {
        var code = 0
        var firstCode = 0
        var firstSymbol = 0
        for count in table.counts {
            code = (code << 1) | (try readBit())
            let index = code - firstCode
            if index >= 0, index < count { return Int(table.symbols[firstSymbol + index]) }
            firstSymbol += count
            firstCode = (firstCode + count) << 1
        }
        throw PayloadPolicyError.malformedFrame
    }

    private mutating func skipBits(_ count: Int) throws {
        for _ in 0..<count { _ = try readBit() }
    }

    private mutating func readBit() throws -> Int {
        if bitCount == 0 {
            guard offset < bytes.count else { throw PayloadPolicyError.malformedFrame }
            bitByte = Int(bytes[offset])
            offset += 1
            if bitByte == 0xff {
                // Never synthesize bits when encountering EOI or a restart marker.
                guard offset < bytes.count, bytes[offset] == 0 else { throw PayloadPolicyError.malformedFrame }
                offset += 1
            }
            bitCount = 8
        }
        bitCount -= 1
        return (bitByte >> bitCount) & 1
    }

    private mutating func finishBits() throws {
        let mask = (1 << bitCount) - 1
        guard bitByte & mask == mask else { throw PayloadPolicyError.malformedFrame }
        bitCount = 0
    }
}
